rust/sbd-client/src/lib.rs

```diff
@@ -0,0 +1,320 @@
+    send_buf: Arc<tokio::sync::Mutex<send_buf::SendBuf>>,
+    read_task: tokio::task::JoinHandle<()>,
+    write_task: tokio::task::JoinHandle<()>,
+}
+
+impl Drop for SbdClient {
+    fn drop(&mut self) {
+        self.read_task.abort();
+        self.write_task.abort();
+    }
+}
+
+impl SbdClient {
+    /// Connect to the remote sbd server.
+    pub async fn connect<C: Crypto>(
+        url: &str,
+        crypto: &C,
+    ) -> Result<(Self, String, PubKey, MsgRecv)> {
+        Self::connect_config(url, crypto, SbdClientConfig::default()).await
+    }
+
+    /// Connect to the remote sbd server.
+    pub async fn connect_config<C: Crypto>(
+        url: &str,
+        crypto: &C,
+        config: SbdClientConfig,
+    ) -> Result<(Self, String, PubKey, MsgRecv)> {
+        use base64::Engine;
+        let full_url = format!(
+            "{url}/{}",
+            base64::engine::general_purpose::URL_SAFE_NO_PAD
+                .encode(crypto.pub_key())
+        );
+
+        let (mut send, mut recv) = raw_client::WsRawConnect {
+            full_url: full_url.clone(),
+            max_message_size: MAX_MSG_SIZE,
+            allow_plain_text: config.allow_plain_text,
+            danger_disable_certificate_check: config
+                .danger_disable_certificate_check,
+        }
+        .connect()
+        .await?;
+
+        let raw_client::Handshake {
+            limit_byte_nanos,
+            limit_idle_millis,
+            bytes_sent,
+        } = raw_client::Handshake::handshake(&mut send, &mut recv, crypto)
+            .await?;
+
+        let send_buf = send_buf::SendBuf::new(
+            send,
+            config.out_buffer_size,
+            (limit_byte_nanos as f64 * 1.1) as u64,
+            std::time::Duration::from_millis((limit_idle_millis / 2) as u64),
+            bytes_sent,
+        );
+        let send_buf = Arc::new(tokio::sync::Mutex::new(send_buf));
+
+        let send_buf2 = send_buf.clone();
+        let (recv_send, recv_recv) = tokio::sync::mpsc::channel(4);
+        let read_task = tokio::task::spawn(async move {
+            while let Ok(data) = recv.recv().await {
+                let data = Msg(data);
+
+                match match data.parse() {
+                    Ok(data) => data,
+                    Err(_) => break,
+                } {
+                    MsgType::Msg { .. } => {
+                        if recv_send.send(data).await.is_err() {
+                            break;
+                        }
+                    }
+                    MsgType::LimitByteNanos(rate) => {
+                        send_buf2
+                            .lock()
+                            .await
+                            .new_rate_limit((rate as f64 * 1.1) as u64);
+                    }
+                    MsgType::LimitIdleMillis(_) => break,
+                    MsgType::AuthReq(_) => break,
+                    MsgType::Ready => (),
+                    MsgType::Unknown => (),
```

Comment:
would this be relevant when we add new message types?

Reply:
Correct. Older clients that don't understand the new message types will just ignore them.